Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt

```diff
@@ -1,42 +1,110 @@
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.flow.transform
+import kotlinx.coroutines.plus
 import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class RecoveryCertificateRepository @Inject constructor(
     @AppScope private val appScope: CoroutineScope,
-    private val dispatcherProvider: DispatcherProvider,
+    dispatcherProvider: DispatcherProvider,
     private val qrCodeExtractor: DccQrCodeExtractor,
-    valueSetsRepository: ValueSetsRepository,
+    private val storage: RecoveryCertificateStorage,
 ) {
 
-    val certificates: Flow<Set<RecoveryCertificateWrapper>> = flowOf(emptySet())
+    private val internalData: HotDataFlow<Set<RecoveryCertificateContainer>> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.IO,
+        sharingBehavior = SharingStarted.Lazily,
+    ) {
+        storage.recoveryCertificates
+            .map { recoveryCertificate ->
+                RecoveryCertificateContainer(
+                    data = recoveryCertificate,
+                    qrCodeExtractor = qrCodeExtractor
+                )
+            }
+            .toSet()
+            .also { Timber.tag(TAG).v("Restored recovery certificate data: %s", it) }
+    }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing data.") }
+            .onEach { recoveryCertificates ->
+                Timber.tag(TAG).v("Recovery Certificate data changed: %s", recoveryCertificates)
+                storage.recoveryCertificates = recoveryCertificates.map { it.data }.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot recovery certificate data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.IO)
+    }
 
+    val certificates: Flow<Set<RecoveryCertificateWrapper>> =
+        internalData.data.transform { set ->
+            set.map { RecoveryCertificateWrapper(null, it) }.toSet()
+        }
+
+    @Throws(InvalidRecoveryCertificateException::class)
     suspend fun registerCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("registerCertificate(qrCode=%s)", qrCode)
-        throw NotImplementedError()
+        return qrCodeExtractor.extract(qrCode.qrCode).toContainer().apply {
+            internalData.updateBlocking {
+                if (any { it.containerId == containerId }) {
+                    throw InvalidRecoveryCertificateException(InvalidHealthCertificateException.ErrorCode.ALREADY_REGISTERED)
+                }
+                plus(this)
+            }
```

Comment:
I think it's the wrong `this`, `this` in that scope might be the current data snapshot from `updateBlocking`?

```kotlin
fun updateBlocking(
    action: suspend Set<RecoveryCertificateContainer>.() -> Set<RecoveryCertificateContainer>
): Set<RecoveryCertificateContainer>
```

`plus(this)` would be set + set, duplicating the original data and ignoring the new value.
So no compiler error because `SET.plus(SET)` is also valid.

You likely meant `this@apply`, but maybe we can write this a little bit different so future devs don't accidentally break this without noticing.

Reply:
put the new container into new var